w3-egress-tracking.md

```diff
@@ -0,0 +1,258 @@
+
+```json
+{  // "/": "bafy..track"
+  "v": "0.9.1",
+  "iss": "did:key:zStorageNode",
+  "aud": "did:web:ETrackerService",
+  "att": [
+    {
+      "can": "space/egress/track",
+      "with": "did:web:ETrackerService",
+      "nb": {
+        "receipts": [
+          "bafy...retrieveRcpt1",
+          "bafy...retrieveRcpt2",
+          ...
+          "bafy...retrieveRcptN"
+        ],
+        "endpoint": "https://storage.node/receipts"
+      }
+    }
+  ]
+}
+```
+
+The retrieval receipts the Storage Node wants to provide are included in the invocation caveats. Instead of attaching the receipts directly to the invocation, only their CIDs are included, along with a URL to fetch them from. The Egress Tracking Service will calculate egress fees based on these receipts, so it is in the best interest of the Storage Node that these receipts are available at the URL provided. Therefore, it is RECOMMENDED that this special receipts endpoint is managed by the Storage Node itself.
+
+### `space/egress/track` receipt example
+
+After processing the invocation, the Egress Tracking Service returns a receipt.
+
+```json
+{
+  "ran": "bafy...track",
+  "out": {
+    "ok": {}
+  },
+  "fx": {
+    "fork": [
+      // Egress tracking records will be consolidated at some point in the future
+      { "/": "bafy...consolidate" }
+    ]
+  }
+}
+```
+Periodically, the Egress Tracking Service (or some other service or component, for that matter) will process tracked egress records sand consolidate them into a view that can be used to calculate egress fees. The effects in the receipt contain a link to a `space/egress/consolidate`, which tells the Storage Node that the egress records will be processed asynchronously. Storage Nodes will be able to fetch receipts of the `space/egress/consolidate` async actions to check the result of the consolidation process.
+
+### `space/egress/consolidate` invocation example
+
+The following example shows a `space/egress/consolidate` invocation sent by the Egress Tracking Service to the Egress Record Consolidator.
+
+```json
+{  // "/": "bafy..consolidate"
+  "iss": "did:key:zETrackerService",
+  "aud": "did:web:ConsolidatorService",
+  "att": [
+    {
+      "can": "space/egress/consolidate",
+      "with": "did:web:ConsolidatorService",
+      "nb": {
+        "cause": { "/": "bafy..track" }
+      }
+    }
+  ]
+}
+```
+
+When receiving a `space/egress/consolidate` invocation, the Egress Record Consolidator will process egress tracking records and produce a receipt. To process egress records, the Egress Record Consolidator will fetch the receipts from the Storage Node using the URL provided in the `space/egress/track` invocation. The data in these receipts, along with the original invocation, will be used to create a view that can be used to calculate egress fees for the Storage Node.
```

Comment:
> The data in these receipts, **along with the original invocation**, will be used to create a view that can be used to calculate egress fees for the Storage Node.

Was searching for this line. The assumption here is that when the egress tracker requests a receipt from a storage node, they are doing so "recursively"? Meaning the expected response from `https://storage.node/receipts/bafy...retrieveRcpt` is a "Full" receipt including the actual invocation that caused it, rather than just a link to it. Is that right?

Reply:
although we usually use links in UCAN invocations and delegations, which means that linked documents could be fetched from a different source, I think all of our implementations assume they'll find the linked blocks attached. 

The intent of that sentence is to specify explicitly that the service expects the invocation to be attached, which we always do right now anyway AFAICT.

Reply:
My comment here just demonstrates my lack of familiarity with how we treat DAGs 😅. 
Previous projects I've worked around would usually just return the DAG node represented by the CID, rather than the entire tree. Getting the whole thing in one request is very convenient here 😄